core/src/main/java/cucumber/runner/Runner.java

```diff
@@ -128,6 +128,9 @@ private void addTestStepsForPickleSteps(List<TestStep> testSteps, PickleEvent pi
                 match = new FailedStepInstantiationMatch(pickleEvent.uri, step, t);
             }
             testSteps.add(new PickleTestStep(pickleEvent.uri, step, match));
+            if (!runtimeOptions.isDryRun()) {
+                addTestStepsForAfterStepHooks(testSteps, pickleEvent.pickle.getTags());
+            }
```

Comment:
Could you provide a test that shows the after hook is executed after each step?

Reply:
This also affects the  `JSONFormatter`. Currently it assumes all hooks are either before or after hooks. Using after step hooks will have undefined results.

Reply:
@brasmusson where do you think the `TestStepStarted` events for step hooks should go in the json? I am thinking they should go into the same list as the the test steps themselves.

Reply:
I can not. It does indeed all seem to work out.

Reply:
What happens when a step fails? Are the after hooks of the following steps skipped?

Reply:
No, right now they aren't skipped. But makes sense to skip them if step is skipped

Reply:
Issue is hooks are marked Unskipable

Reply:
The code emulates invoke-around semantics but doesn't fully implement the concept.

Reply:
what do you mean by that?

Reply:
what part of the implementation is missing?

Reply:
Right now cucumber represents execution as a series of steps. For completeness I'll assume there is also a before step hook. It might look like this:

```
before hook
beforeStep hook
step1
afterStep hook
beforeStep hook
step2
afterStep hook
after hook
```

What it should try to achieve is:

```js
executeScenario(){
    executeBeforeHooks()
    executeSteps()
    executeAfterHooks()
}

executeSteps(){
    skipNextStep = false;

    for step in steps
      skip = skipNextStep
      beforeStepHook(skip)
      skipNextStep = step(skip)
      afterStepHook(skip)
      skip = skipNextStep
}
```

What it currently does in TestCase:
```
executeSteps(){
    skipNextStep = false;

    for step in steps
      skipNextStep = step(skipNextStep)
}
```

So not only should after step hooks be skipabble, they should also only be skipped when their step was skipped.

Reply:
I got your point. Reason for not implementing BeforeStep is I do not see any use case for the same. If you see previous step-hook requests, they all requested for AfterStep hook which I also feel is useful in many scenarios.
I do not have an issue implementing BeforeStep, my only point is it should not be implemented just because we are implementing AfterStep

Coming to skipable hook, currently all Hooks are of class *UnskipableStep*, if I workaround that, it defeats the purpose of UnskipableStep class. Other option is to have class called *SkipableHookStep* for AfterStep (and if required, BeforeStep) hooks.

The logic you mentioned for skipping has a catch, Any failed step will set skipNextStep to true, But failed step should not skip its AfterStep. AfterStep should only be skipped for skipped Steps, not failed Step. I can see how that can be done.

Any thoughts on having a new Class *SkipableHookStep* for AfterStep/BeforeStep?

Reply:
Another option is to rename class **UnskipableStep** to **HookStep** and then internally handle whether to execute hook step or not in **execute()** method

Reply:
Cucumber-Ruby does add AfterStep hooks as ordinary steps, so the AfterStep hooks for failed steps are skipped. This may not be ideal, but on the other hand inconsistency between Cucumber implementation is not ideal either, so we should try to get a common vision on where all the Cucumber implementation are heading.

Reply:
>  I can see afterstep being reported properly in json report

I do not see any test case that specify how after steps hooks are reported in the json report, so I think such a comment is premature.

In Cucumber-Ruby the AfterStep hooks are included in the json report in the "after" entry of each step, see https://github.com/cucumber/cucumber-ruby/blob/master/spec/cucumber/formatter/json_spec.rb#L671-L683

Reply:
I will update the class to HookStep instead of Unskipable step and handle the logic of skipping AfterStep for a failed/skipped step. This will be in sync with ruby.

Let me know if that should be ok

Reply:
I will have to do some changes for JSONFormatter as there is some incosistency when compared with Ruby Json report. Will add a test for same

Reply:
@mpkorstanje , @brasmusson , I have implemented logic of skipping AfterStep in case of failed or skipped Step and also tests for JSON formatter showing AfterSteps JSON arry node should belong to it's respective step node.

Would you mind reviewing the changes?

Reply:
I'll look into them, but it will probably take half a week or so before I have the time and energy to do it.

Reply:
Sure, Thanks